strip whitespaces in render_block_core_cover

navigatrum opened the pull request:
<!-- Thanks for contributing to Gutenberg! Please follow the Gutenberg Contributing Guidelines:
https://github.com/WordPress/gutenberg/blob/trunk/CONTRIBUTING.md -->

## What?
Follows on from [#39658]( https://github.com/WordPress/gutenberg/pull/39658) , https://github.com/WordPress/gutenberg/pull/39658#issuecomment-1117870716
Remove whitespaces from html tags generated by `render_block_core_covermarkup` before injecting the featured image.

## Why?
The fetaured image isn't injected from a template file beacuse of white spaces between html tags (https://github.com/WordPress/gutenberg/pull/39658#issuecomment-1118514002)

navigatrum committed:
strip whitespaces in render_block_core_cover

Follows on from [#39658]( https://github.com/WordPress/gutenberg/pull/39658)
Remove whitespaces from html tags generated by `render_block_core_covermarkup` before injecting the featured image

draganescu commented on packages/block-library/src/cover/index.php:
```diff
@@ -63,7 +63,7 @@ class="wp-block-cover__image-background"
 			'</span><div',
 			'</span>' . $image . '<div',
-			$content
+			preg_replace( '/>(\s)+</m', '><', $content )
```
Can we make the regexp only target the close span- open div markup?

navigatrum replied:
(Sorry for the late reply)
Yes, this is working:
```php		
$content = str_replace(
    '</span><div',
    '</span>' . $image . '<div',
    preg_replace( '/<\/span>(\s+)?<div/', '</span><div', $content, 1 )
);
```
Edit: sorry I wrote `$content = preg_replace(...`, by the way it'is: `$content = str_replace(...`

draganescu replied:
I meant use something like `preg_replace( '/<\/span>(\s)+<div/m', '/span><div', $content )` to be focused on this particular thing we're replacing not the entire block html.

navigatrum replied:
> I meant use something like `preg_replace( '/<\/span>(\s)+<div/m', '/span><div', $content )` to be focused on this particular thing we're replacing not the entire block html.

I understood 😄. And I was suggesting : ` preg_replace( '/<\/span>(\s+)?<div/', '</span><div', $content, 1 )`.
The `multiline` flag is not required. Also, we should use `(\s+)?` instead of  `(\s)+` to ensure that things work also when there are no characers between `</span>` and `<div/`. And add the forth parameter to force exit after the first replacement.

draganescu replied:
Excellent! Can you add a commit with your proposed regexp?

navigatrum replied:
Done! If you prefer a new PR let me know

draganescu reviewed (dismissed):
Thanks for taking care of this bug! :) I assumed the markup in the block's save function is fixed. I'd make the regexp more specific as ideally in the future we'd want to not need it at all so not encouraging reliance on it would be good.

@dmsnell this is spot on about our conversation around hand editing block markup and the assumptions we can make around that. It didn't break the editor in this case but it did break the expected idea that the final output is decided on save.

dmsnell commented:
> It didn't break the editor in this case but it did break the expected idea that the final output is decided on save.

Are you talking about the rendered page @draganescu? We shouldn't expect that the serialized HTML produced by Gutenberg and stored in `post_content` is the HTML someone will retrieve when reading a page online. Server-side processing is fundamental to how WordPress and Gutenberg interact.

We want to remember that the stored HTML is a serialized version of the document, but that [Gutenberg posts themselves are not inherently HTML](https://fluffyandflakey.blog/2017/09/04/gutenberg-posts-arent-html/) and the HTML they generate is both a stored serialization of those posts _and_ a mechanism to provide more graceful fallback for preserving content when the supporting systems (like WordPress) are unavailable.

ndiego commented:
Hi all, is this still something we are trying to get into 6.0, or can we push it to 6.0.1? Just let me know and I will update the project boards accordingly.

ndiego commented:
Given that we are hours from 6.0 RC3, I am going to bump this to the [6.0.1 Project Board](https://github.com/orgs/WordPress/projects/31).

navigatrum committed:
amend regex

dmsnell commented:
At first when I saw this I thought there was a very specific reason we needed the whitespace to be gone, but after looking a bit deeper it looks like we really just need to ensure that we can inject the image into the template.

Did you consider doing real parsing of the HTML in order to inject the image instead of relying on this considerably-more-frail text-parsing? I'm not sure what exactly your needs are, but I find `replace( </span><div>, </span><img><div> )` to be incredibly confusing as a reader: what's the expected markup? what's the goal of the change? why these unlabeled tags? In fact, I don't really know how to review this because of that confusion. I'm going to assume we want to insert the featured image right before the `wp-block-cover__inner-container` `<div>` opens?

<details><summary>Parsing HTML</summary>

```php
function render_block_core_cover( $attributes, $content ) {
	if ( 'image' !== $attributes['backgroundType'] || false === $attributes['useFeaturedImage'] ) {
		return $content;
	}

	if ( ! ( $attributes['hasParallax'] || $attributes['isRepeated'] ) ) {
		$attr = array(
			'class'           => 'wp-block-cover__image-background',
			'data-object-fit' => 'cover',
		);

		if ( isset( $attributes['focalPoint'] ) ) {
			$object_position              = round( $attributes['focalPoint']['x'] * 100 ) . '%' . ' ' . round( $attributes['focalPoint']['y'] * 100 ) . '%';
			$attr['data-object-position'] = $object_position;
			$attr['style']                = 'object-position: ' . $object_position;
		}

		$image = get_the_post_thumbnail( null, 'post-thumbnail', $attr );
		if ( ! $image ) {
			return $content;
		}

		$dom = new DOMDocument();
		$dom->loadHTML( $content );

		$inner_container = null;
		foreach ( $dom->getElementsByTagName('div') as $node ) {
			if ( 1 === preg_match( '/(^| )wp-block-cover__inner-container( |$)/', $node->getAttribute('class') ) ) {
				$inner_container = $node;
				break;
			}
		}

		if ( ! $inner_container ) {
			return $content;
		}

		$image_node = $dom->createDocumentFragment();
		$image_node->appendXML( $image );
		$inner_container->insertBefore( $image_node );

		$content = '';
		foreach ( $dom->documentElement->firstChild->childNodes as $node ) {
			$content .= $dom->saveHTML( $node ) . PHP_EOL;
		}
	} else {
		if ( in_the_loop() ) {
			update_post_thumbnail_cache();
		}
		$current_featured_image = get_the_post_thumbnail_url();
		if ( ! $current_featured_image ) {
			return $content;
		}

		$dom = new DOMDocument();
		$dom->loadHTML( $content );

		$container = null;
		foreach ( $dom->getElementsByTagName( 'div' ) as $node ) {
			if ( 1 === preg_match( '/(^| )wp-block-cover( |$)/', $node->getAttribute( 'class' ) ) ) {
				$container = $node;
				break;
			}
		}

		if ( ! $container ) {
			return $content;
		}

		$background_style = sprintf( 'background-image:url(%s)', $current_featured_image );
		$container->setAttribute('style', $container->getAttribute('style') . ' ' . $background_style );

		$content = '';
		foreach ( $dom->documentElement->firstChild->childNodes as $node ) {
			$content .= $dom->saveHTML( $node ) . PHP_EOL;
		}
	}

	return $content;
}
```
</details>

^^^ that's a little convoluted, but a few helper functions can make it read much smoother

<details><summary>Parsing HTML with helpers</summary>

```php
function render_block_core_cover( $attributes, $content ) {
	if ( 'image' !== $attributes['backgroundType'] || false === $attributes['useFeaturedImage'] ) {
		return $content;
	}

	function html2dom( $html ) {
		$dom = new DOMDocument();
		$dom->loadHTML( $html );

		return $dom;
	}

	function dom2html( $dom ) {
		$html = '';
		foreach ( $dom->documentElement->firstChild->childNodes as $node ) {
			$html .= $dom->saveHTML( $node ) . PHP_EOL;
		}

		return $html;
	}

	function html2node( $dom, $html ) {
		$fragment = $dom->createDocumentFragment();
		$fragment->appendXML( $html );

		return $fragment;
	}

	function getTagWithClassname( $dom, $tag, $class ) {
		foreach ( $dom->getElementsByTagName( $tag ) as $element ) {
			if ( 1 === preg_match( '/(^| )' . preg_quote( $class, '/' ) . '( |$)/', $element->getAttribute('class') ) ) {
				return $element;
			}
		}

		return null;
	}

	if ( ! ( $attributes['hasParallax'] || $attributes['isRepeated'] ) ) {
		$attr = array(
			'class'           => 'wp-block-cover__image-background',
			'data-object-fit' => 'cover',
		);

		if ( isset( $attributes['focalPoint'] ) ) {
			$object_position              = round( $attributes['focalPoint']['x'] * 100 ) . '%' . ' ' . round( $attributes['focalPoint']['y'] * 100 ) . '%';
			$attr['data-object-position'] = $object_position;
			$attr['style']                = 'object-position: ' . $object_position;
		}

		$image = get_the_post_thumbnail( null, 'post-thumbnail', $attr );
		if ( ! $image ) {
			return $content;
		}

		$dom = html2dom( $content );
		$inner_container = getTagWithClassname( $dom, 'div', 'wp-block-cover__inner-container' );
		if ( ! $inner_container ) {
			return $content;
		}

		$inner_container->insertBefore( html2node( $dom, $image ) );
		return dom2html( $dom );
	}

	if ( in_the_loop() ) {
		update_post_thumbnail_cache();
	}
	$current_featured_image = get_the_post_thumbnail_url();
	if ( ! $current_featured_image ) {
		return $content;
	}

	$dom = html2dom( $content );
	$container = getTagWithClassname( $dom, 'div', 'wp-block-cover' );
	if ( ! $container ) {
		return $content;
	}

	$background_style = sprintf( 'background-image:url(%s)', $current_featured_image );
	$container->setAttribute( 'style', $container->getAttribute('style') . ' ' . $background_style );

	return dom2html( $dom );
}

```
</details>

Looks like this is now outdated based on a refactor of the function as well; we need to update the patch to account for that. I've made my suggestion here based on the more recent version of the file in `trunk`. Also I added a couple adjustments to handle cases when there's no thumbnail etc… Note that here we're appending the style to the `style` attribute instead of adding a new `style` attribute. According to my tests if we add multiple `style` attributes to an Element then only the last will apply; in other words, we don't merge styles, we only take the last available one.

The most relevant bit is this code here as it's the part that disambiguates the replacement operation (at least in my head it does a better job indicating what _should_ be happening) and its resilient to whitespace, escaping, and formatting issues. As a bonus, it gives us the classnames as complete strings we can find with a basic search when we need to find and update this code:

```php
$dom = html2dom( $content );
$inner_container = getTagWithClassname( $dom, 'div', 'wp-block-cover__inner-container' );
if ( ! $inner_container ) {
	return $content;
}

$inner_container->insertBefore( html2node( $dom, $image ) );
return dom2html( $dom );
```

navigatrum commented:
@dmsnell  I had considered the idea of Dom parsing, but I discarted it because because:
1) @draganescu didn't use it
2) `DOMDocument` is used in Wordpress very rarely compared to its potential fields of application, and not rarely with some disclaimers and after checks. See:
- > [we cannot universally count on DOMDocument being available](https://github.com/WordPress/wordpress-develop/blob/edcfb250d3bfffb86ffb5e8e784864b2b852b2c5/src/wp-includes/class-wp-simplepie-sanitize-kses.php#L14)
- [`if ( ! class_exists( 'DOMDocument', false ) )`](https://github.com/WordPress/wordpress-develop/blob/5a49274a927b21ce18165bcba004c20421fcc088/src/wp-admin/includes/misc.php#L816)
- >Handling HTML on the server
As hinted in § Background, PHP can deal with HTML, but it’s a delicate operation. It’s no accident that most of WordPress has traditionally handled content as strings throughout its components: the tradeoff is less power in exchange for performance and safety. Thus, our utmost diligence is required when we set out to introduce a system that processes a new HTML document on every render_block. (from [#26111])

draganescu commented:
@dmsnell I think @navigatrum only applied a simple correction for code that was already merged, so their approach is valid in the context of how this already works. 

While your suggestion to use DOMDocument to build block HTML on render is **very good**, it is something we as a project have not made use of so far (e.g. `gutenberg_restore_group_inner_container`, `render_block_core_image`, `render_block_core_post_featured_image`, `render_block_core_template_part`, `render_block_core_site_logo`). 

From what I see, in general block HTML is generated with string manipulation. While what we do in the cover block, for the featured image case, is "injecting" html into whatever the block is saved as, is a bit different than general rendering, using DOMDocument to build a DOM object would be a completely new way of achieving a similar result. 

The problem this particular PR tries to resolve is that manually editing block html in theme files may insert spurious whitespace. It is a general block problem that we happened to stumble upon here.

cc @hellofromtonya - should we try to move away from HTML string manipulation and favour DOM building? Are the old considerations that @navigatrum mentioned still a thing?

hellofromtonya commented:
>cc @hellofromtonya - should we try to move away from HTML string manipulation and favour DOM building? Are the old considerations that @navigatrum mentioned still a thing?

@draganescu No, I don't recommend building with [PHP's DOM API](https://www.php.net/manual/en/book.dom.php). Why? 

PHP's DOM API:
- the API can be disabled by web hosts
- [requires the `libxml` extension](https://www.php.net/manual/en/dom.requirements.php) (min version of 2.6.0) which may or may not be available on all web hosts 
- struggles to handle malformed HTML. For example, a missing closing tag such as `</div>` or `</head>` can be impossible for it to detect and resolve
   - if malformed and it can't resolve it, then modifying nodes and eventually saving it out as HTML could mean broken layouts
- encoding (other than UTF-8) can be problematic

I'd love to see a DOM API land in Core. But, as noted above, when it can't work/run, then string fallbacks are still needed. Until the above are resolved, the use of PHP's DOM API is too limited.

dmsnell commented:
:sigh:

earlier today I spent a while writing a large response and now I'm not seeing the comment _I'm sure I posted_. I'm going to try and write most of it again, but I'm sorry I won't be able to do it as eloquently or completely as I did before.

@hellofromtonya do we have any data on how many WordPress hosts have `DOMDocument` disabled? there's at least a couple un-guarded uses of it in Core already and I'm wondering if we have WordPress crashing in parts of the web we're unaware of.

@youknowriad @gziolo have you encountered this before? I've been considering more broad use of `DOMDocument` in Gutenberg to avoid a lot of common mistakes with over-simplifying HTML matching and parsing.

@navigatrum @draganescu it would be nice to consider the other suggestions I had, specifically in sharing more context for what we're changing and searching for. I'm trying to put my feet into the shoes of someone tasked with figuring out why certain markup is changing and broken and particularly the lack of anything in there that I can put into a simple string search and find this code is a big obstacle to debugging it.

alternatively we can make the RegExp pattern a little more verbose and get a lot more safety in the process. although this looks more complicated at the time of writing, I'm thinking it would be far more resilient in time and more helpful to the future given that it communicates from the code the specific thing it's looking for.

Fun fact: in PHP we are free to use other pattern delimiters which means we don't have to escape `/`

```php
$inner_container_start = '~<div [^>]*\bclass=("|')((?!\1).)*\bwp-block-cover__inner-container\b((?!\1).)*\1( |>)~';

if ( 1 !== preg_match( $inner_container_start, $content, $matches, PREG_OFFSET_CAPTURE ) {
    // handle this, we can't find our div, maybe another plugin changed it
    return $content;
}

list( $fullmatch, $offset ) = $matches[0];
$content = substr( $content, 0, $offset ) . $image_tag . substr( $content, $offset );
```

this too can be split into functions that give better names to what we're looking for if it helps make the searching and replacement clearer.

```php
function strpos_tag_with_class( $html, $class ) {
    $CLASS = preg_quote( $class, "~" );
    $pattern = "~<div [^>]*\bclass=(\"|')((?!\1).)*\b{$CLASS}\b((?!\1).)*\1( |>)~";

    if ( 1 !== preg_match( $pattern, $html, $matches, PREG_OFFSET_CAPTURE ) ) {
        return false;
    }

    list( $fullmatch, $offset ) = $matches[0];
    return $offset;
}

…

$image_at = strpos_tag_with_class( $content, "wp-block-cover__inner-container" );
if ( false === $image_at ) {
   return $content;
}

$before_image = substr( $content, 0, $image_at );
$after_image = substr( $content, $image_at );

return "{$before_image}{$image_tag}{$after_image}";
```

youknowriad commented:
I think regex have been favored in general for performance reasons (but I may wrong), especially since all this logic renders in the frontend and we have a number of similar use-cases. 

I wouldn't mind DOMDocument usage if it behaves better in terms of performance.

hellofromtonya commented:
>do we have any data on how many WordPress hosts have `DOMDocument` disabled? 

I'm not sure that info is collected. @dd32 might know.

>there's at least a couple un-guarded uses of it in Core already and I'm wondering if we have WordPress crashing in parts of the web we're unaware of.

@dmsnell besides build tooling and tests, can you point me to where `DOMDocument` is unguarded in WP Core? Thank you :) 

Doing a search in the 6.0 released source (in build rather than develop), `DOMDocument` is used for (a) XML (e.g. SimplePie, `WP_oEmbed`, IIS web config files) and (b) detection - not to modify HTML nodes - (e.g. detect legacy in `WP_Widget_Text`, `iis7_supports_permalinks()` ). I'm not seeing unguarded usage.

hellofromtonya commented:
Thinking more about this, I think I need to provide more clarity of where `DOMDocument` could be considered. 

Possible use cases:

- XML
- Detection (with fallback)
- Possibly node selection (see <sup>*</sup> note below)

<sup>*</sup>Possibly node selection: fallback would be needed if `DOMDocument` or `libxml` not available; and if selecting with translated content, then extensive testing and feedback especially with non-UTF-8 languages to ensure translated content renders correctly (encoding).

If there's zero risk of malformed HTML being passed into `DOMDocument`, e.g. the HTML structure is fully controlled within a block, and node selection does not include translatable content (meaning no encoding issues), then `DOMDocument` might be an appropriate tool. A fallback is needed for those users whose hosts do not have the libraries.

For these use cases, a Core DOM API might be handy to pull all the library checks, setups (such as turning on and off libxml errors, encoding, etc.), error handling, fallback detection, etc. into one place vs. distributing repeating/redundant code.

navigatrum commented:
>If there's zero risk of malformed HTML being passed into DOMDocument, e.g. the HTML is controlled within a block

@hellofromtonya Problem is when a block 'becomes' part of a theme template, HTML tends to become malformed. In my case I just copyed the code from the Gutenberg code Editor, and pasted it to my code editor. The auto formatter did the rest.

draganescu commented:
@navigatrum just wanted to note that this tiny PR here takes so long not because it's wrong, but because it raises some possibly important conclusions around how we expect the system to be both resilient to hand coding markup and handling it's output consistently.

The replacement of spaces solution is good, but should we do that everywhere? Are there other blocks where a simple space in the template markup breaks the system? Hence the duration. Wanted to mention that and to thank you for the patience 🥇

draganescu commented:
@navigatrum I think for this particular PR, we can just make it so that we don't introduce code that will stump us in the future via @dmsnell 's suggestions:

- let's add a comment explaining why the space removal is done
- let's make the expression more verbose, maybe include the block classname, so that it becomes more clear where in the markup the img is inserted and what is the expected shape of that markup.

What do y'all think?

anton-vlasenko commented:
I used `DOMDocument` sometime in 2008, so I'm not an expert on the subject.
But I would rather not use it.
As noted [above](https://github.com/WordPress/gutenberg/pull/40859#issuecomment-1134854433), WordPress sites will have to depend on libxml and the DOM library, and they are not available everywhere. Therefore, there must be a fallback mechanism (and if there is such a mechanism, then there is no need to depend on the DOMDocument class :).
The issue with handling invalid HTML code can be solved by using the  [LIBXML_HTML_NOIMPLIED](https://www.php.net/manual/en/libxml.constants.php#constant.libxml-html-noimplied) option, but there is no guarantee that it will work correctly in all cases.
I favor using regular expressions and/or solving this issue in JavaScript (if possible).

dmsnell commented:
> I think regex have been favored in general for performance reasons (but I may wrong), especially since all this logic renders in the frontend and we have a number of similar use-cases.
> I wouldn't mind DOMDocument usage if it behaves better in terms of performance.

@youknowriad this is the PHP `DOMDocument` we're talking about and all this is done in the `render_block` family of hooks.

@hellofromtonya upon further inspection I found that use in an expired `build` directory of Gutenberg but didn't notice it's stale because that directory is included in the `.gitignore`, so it looked to me like a real file. I'm not seeing unguarded uses after all

azaozz commented:
Late to the party but may I add 2c? :) 

DOMDoculemt sucks for parsing HTML. Enough said. It's great for generating (simple) XML. It works pretty well for parsing well-formed XML. But `libxml` cannot handle parsing HTML like a browser, and afaik it is not meant to. The main difference is that the browsers do a lot of "fixing" of badly formed or invalid HTML (including things like character encodings and entities, missing or extra closing tags, outdated/invalid stuff, and many more).

@dmsnell I think we had a similar conversation few years ago, tried using DOMDocument then too, and failed.

At the same time using Regex for "partially parsing" HTML sucks too. Core already has some examples there. If a tag has to be injected and that cannot be done in the browser, best would be to append (or prepend) the tag to the rest of the HTML in the block. That would probably look more like:

```
if ( '</div>' === substr( $block_html, -6 ) ) {
    $block_html = substr( $block_html, strlen( $block_html ) - 6 );
    $block_html = $block_html  . $my_img_tag . '</div>';
}
```

The rest (positioning, etc.) can be handled with CSS.

dmsnell commented:
> The replacement of spaces solution is good, but should we do that everywhere? Are there other blocks where a simple space in the template markup breaks the system?

Overall I think this is a disturbing indicator and ultimately falls back to the use of string-processing for HTML content. No whitespace changes should break HTML (apart from the fact that a leading newline/space after the opening of a tag inserts a single whitespace on render).

If we get to the point where we find that whitespace changes did break our code a far more reliable fix would be to update our code that depends on that whitespace sensitivity. Gutenberg itself is liberal in reformatting HTML so a quick hack here and there to force a specific HTML formatting so that processing stages later down the line don't crack is going to build a mess -quickly-, not to mention the readability issues when finding code like `replace( '~</span>\s+<div~', '</span><div' )` (which is incredibly confusing and somewhat frustrating to come across).

I'd rather see us frame this as "the replacement of spaces solution _works in this particular situation_ but we should try to eliminate the need for such a workaround that we'll soon regret."

Realistically this isn't the place or PR to demand that everything properly parse HTML (whether through `DOMDocument()` or through very complicated recursive PCRE patterns - yes PCRE can adequately parse HTML because they are more powerful than basic "regular expressions"). Still, we can get a lot better practically speaking and not have to come back here repeatedly as this crops up again.

In other words, the problem isn't the space before the `<div>` - the problem is that we're using too much of an over-simplified operation to inject the image tag and if we improve our matching code _just a little bit_ we won't even have to talk about the whitespace 😄

azaozz commented:
> I think this is a disturbing indicator and ultimately falls back to the use of string-processing for HTML content.

Yeah, completely agree. String processing of HTML is best avoided at all costs. It may look simple and bulletproof but always has "strings attached" (pun intended) :)

dmsnell commented:
@azaozz I missed your comment while I was writing mine.

> I think we had a similar conversation few years ago, tried using DOMDocument then too, and failed.

sounds familiar: image source sets? I was hoping things have changed since then.

> At the same time using Regex for "partially parsing" HTML sucks too.

Not sure I fully understand what you're saying, but I believe the code snippet you wrote is roughly equivalent to the one I wrote, and the one I wrote I think has the same failure modes, where this HTML appears inside a `CDATA` section. given that I couldn't figure out what the intentional change was, I assumed it was specific to the `<div>` with the particular class name so wanted to match on that. of note, I think we're injecting the image tag _before_ the `<div>` starts, not at the end of it.

azaozz commented:
> sounds familiar: image source sets?

Yep, think so! :) 

> the code snippet you wrote is roughly equivalent to the one I wrote

Yes, it is. Was just trying to point out that it would be best to append (or prepend) an img tag at the end (or the beginning) of the block's HTML, not inject it in the middle. Then it would be a simple string concatenation, won't need regex, etc. Everything else can (most likely) be handled with CSS.

navigatrum commented:
@draganescu 
(You are welcome, I'm happy to give a hand)
1) Looking at: https://github.com/WordPress/gutenberg/blob/51db4bf888e6b18cf9d18266108114d61070f3ad/lib/block-supports/layout.php#L232
1.1) since here the class is matched with just `\b[^>]*wp-block-group`, I'd propend to skip the `class=(\"|')((?!\1).)*` part of the regex by [dmsnell](https://github.com/dmsnell), i.e.: `~<div [^>]*\bclass=("|')((?!\1).)*\bwp-block-cover__inner-container\b((?!\1).)*\1( |>)~`, which is more precise, however.
1.2) I'd suggest, instead, to identify our famous `span` before our infamous `div`.
So what about:
    ```php
    /* Inserts the featured image between the (1st) cover 'background' `span` and 'inner_container' `div`,
       and removes eventual withespace characters between the two (typically introduced at template level) */
    $content = preg_replace(
	    '/(<span\b[^>]*wp-block-cover__background[\s|"][^>]*>.*<\/span>)\s*(<div\b[^>]*wp-block-cover__inner-container[\s|"][^>]*>)/U',
	    "$1$image$2",
	    $content,
	    1
    );
    ```

2) I would tend to limit variables used just one time as much as possible in production code. But if you really prefer sometihng like:
    ```php
    /* Comment */
    $background_and_container_regex = // regex
    $contenet = preg_replace($background_and_container_regex, "$1$image$2", $content, 1);
    ```
    let me know (again 😄 )

3) Not sure about the `>.*</span>` part of my regex. Should we force `></span>`?

dd32 commented:
> > do we have any data on how many WordPress hosts have `DOMDocument` disabled?
> 
> I'm not sure that info is collected. dd32 might know.

Can confirm: We don't have any stats on it. https://core.trac.wordpress.org/ticket/48116 would be needed for that.

As @azaozz said:
> DOMDoculemt sucks for parsing HTML. 

Can confirm this too! It works amazing for many basic use-cases, but as soon as you hit one of it's edge-cases in parsing you'll wish you used regex..

draganescu commented:
@navigatrum I think what you suggest at (1) in https://github.com/WordPress/gutenberg/pull/40859#issuecomment-1138121109 is fine to add as a commit and then we can see how it works. 🙇🏻‍♂️

navigatrum committed:
Cover-block: edit regex to insert the featured img

For the block layout see: https://github.com/WordPress/gutenberg/blob/trunk/packages/block-library/src/cover/edit/index.js#L327-L367
For the HTML tags capturing by class name see: https://github.com/WordPress/gutenberg/blob/51db4bf888e6b18cf9d18266108114d61070f3ad/lib/block-supports/layout.php#L232

azaozz commented on packages/block-library/src/cover/index.php:
```diff
@@ -60,10 +60,13 @@ class="wp-block-cover__image-background"
-			'</span><div',
-			'</span>' . $image . '<div',
-			$content
+		/* Inserts the featured image between the (1st) cover 'background' `span` and 'inner_container' `div`,
```
Why does the featured image have to be inserted there? Who decided that and for what purpose? Could you please comment to make sure we're on the same page (I'm probably missing some considerations there).

navigatrum replied:
Probably you didn't read the first line of comment to this PR.
It says:
>For the block layout see: https://github.com/WordPress/gutenberg/blob/trunk/packages/block-library/src/cover/edit/index.js#L327-L367

Fell free to take a look at the history of that file: the *who* and their *purpose* cannot be really hidden. My 2¢? A henchman of [his](https://ma.tt/). 😄

draganescu replied:
@azaozz I decided that for not having a better way - as Gutenberg is missing a token system - to make the cover block get the featured image at render time.

If you look at the [history](https://github.com/WordPress/gutenberg/pull/39658) of PRs in which [I tried](https://github.com/WordPress/gutenberg/pull/39515) to [solve](https://github.com/WordPress/gutenberg/pull/39410) the featured image in cover block problem you'll see there are a few constraints that arise from our lack of a token system which results in this kind of render string operations on the block markup.

It is not the only block where these operations occur and I think it also happens  (e.g. `gutenberg_restore_group_inner_container`, `render_block_core_image`, `render_block_core_post_featured_image`, `render_block_core_template_part`, `render_block_core_site_logo`).

azaozz replied:
Right, my bad, should have explained this better. Was thinking that the way this works may have been simpler if the featured image was appended to the end of the block's HTML. The difference in the featured image's node position would have probably been pretty easy to account for with a little of CSS. Even maybe that could have been preferred in some cases as would have made it possible to determine when an image was inserted there.

azaozz commented on packages/block-library/src/cover/index.php:
```diff
@@ -60,10 +60,13 @@ class="wp-block-cover__image-background"
+		/* Inserts the featured image between the (1st) cover 'background' `span` and 'inner_container' `div`,
+		   and removes eventual withespace characters between the two (typically introduced at template level) */
+		$content = preg_replace(
+		    '/(<span\b[^>]*wp-block-cover__background[\s|"][^>]*>.*<\/span>)\s*(<div\b[^>]*wp-block-cover__inner-container[\s|"][^>]*>)/U',
```
"Processing" HTML with regex sucks no matter how you look at it. The more complex the regex, the more it sucks :) 

This looks very fragile. Is there any chance at all to do it in a more compatible and stable manner? One thing that comes to mind is to prepend or append the `<img>` tag, not attempt to insert it in the middle of some HTML. Trying to "splice" it as the current code does would most likely keep introducing edge cases, will likely break if the block is ever changed by a plugin, and will require ongoing maintenance.

If prepending/appending is impossible, seems there are other options:
1. Insert an HTML element as a placeholder in the editor. That way the specific location of the featured image can be determined much better, and the replacement code on the PHP side will be a bit more "bulletproof". (The placeholder can be removed when rendering the block and there's no featured image.)
2. Reconstruct the block's HTML in PHP, i.e. make it really a dynamic block.

dmsnell replied:
One note on this regex: there's a major jump in required complexity between matching an HTML tag and parsing HTML. Parsing HTML tags is simpler and possible with traditional RegExp patterns and knowledge, but what we've done by attempting to capture the `<span>` and the `<div>` together is parsing full HTML. A trivial example input shows how [this can break down](https://regex101.com/r/VHvK8B/1):

```
<span class="wp-block-cover__background"><span class="wp-block-cover__background">Something else</span><div class="wp-block-cover__inner-container">Something</span><div class="wp-block-cover__inner-container">
```

The `.*` really is problematic, and you probably don't want to have to maintain an appropriate RegExp pattern to do this properly. There's no good way to match that specific `</span>` with a RegExp, which is why in my example I matched the `<div>` with its class name, because that's very doable. That is, if we are trying to identify a specific tag based on some attribute, we can only do that by matching the opening tag (or a void tag if there are no children). If we want to match the _end_ of a tag we only have a good option to match "any closing tag of this kind."

Also curious, why do we want to drop the match on the `class` attribute?

navigatrum replied:
@dmsnell 
1. Abut your curisity, take a look: https://github.com/WordPress/gutenberg/blob/51db4bf888e6b18cf9d18266108114d61070f3ad/lib/block-supports/layout.php#L220-L234
    
    As you can see, no `class=` to match the div with `class="wp-block-group"`, and the div with`class="wp-block-group__inner-container"` 
    
    And again: https://github.com/WordPress/gutenberg/blob/51db4bf888e6b18cf9d18266108114d61070f3ad/lib/block-supports/layout.php#L243-L246
    
    No `class=`. And the matching of a closing tag. And `.*` between the two.
    
    As in our case, they had to put something between the two groups.

2. With your **HTML malformed** every day usecase, putting a `div` inside a `span`:
  ```html
  <span><span>...</span><div>...</span><div>...
  ``` 
  my regex would insert the featured image before the last `div`. While the regex you suggested [before the first `div`](https://regex101.com/r/VHvK8B/1).
    So what?

It would seem you are maintaining at the same time:
- "It's a bad practice trying to cotrol what comes before the opening div. That's why I was suggesting no control at all"
- "I can't accept we are so permissive about what comes (within the span) before the div.We must absolutly control it"

navigatrum replied:
>0) One thing that comes to mind is to prepend or append the <img> tag, not attempt to insert it in the middle of some HTML.

I could be wrong, but I bet it could accomplished rewriting less than 10% of the entire Gutenberg!

> 1) Insert an HTML element as a placeholder in the editor.

Easy: rewite the Cover Block (and probably the Image Block)!

2) > Reconstruct the block's HTML in PHP

More or less this is what has been done by @draganescu. Who has added such a cool feature to WP. I was so impacient that installed Gutenberg for the first time to try it. And I had the misfortune to discover some kind of bug. It's not a real bug. And it wasn't me to discover the nature of this pseudo 'bug'. We have to remove some spaces and line breaks. That's it. If yu have some code, let's do it!

dmsnell replied:
@navigatrum first of all I want to say thank for working hard on this. please know that when I look at a proposed change my purpose and goal is to try and find the places where it could break down and so if I seem pessimistic it's because that's how I'm looking at the changes (and all changes for any PR).

Concerning the `class` question would it be right to understand that you are saying you removed it because there was other code in the project that also doesn't? If so, I wouldn't suggest we need to feel bound to match that. I'm not sure why it was left out there either, and it's not a critical thing, but it just seems surprising that we'd want to intentionally decouple the class name we're looking for from the `class` attribute we expect to find it in. 🤷‍♂️ 

> my regex would insert the featured image before the last div. While the regex you suggested [before the first div](https://regex101.com/r/VHvK8B/1).
So what?

I might not have presented a great example because it looks malformed. In [this example](https://regex101.com/r/LCAVkP/1) it should be more apparent that the `.*` will continue consuming input as long as it can while still matching the entire thing. Even a small change there could prevent a match from continuing past any unexpected tags (for example, if we use `[^<]*` instead of `.*` then we at least shouldn't break past unexpected tags that might sit between the `<span>` and the `<div>`).

We're probably somewhat shielded here because we're only being supplied the HTML from within the block itself (and also because we added the match on the class name), but if this were applied to the post (or if we were only looking at the tag names) we could easily get into trouble.

> It would seem you are maintaining at the same time:

I think there's a misunderstand here. What I'm saying is that we should be aware that whitespace isn't significant in HTML in the way we were treating it and we should look for an approach that doesn't depend on a specific whitespace formatting. We have very few guarantees about how it will reach our code.

When moving to the more robust RegExp there was simply another point I wanted to raise about the risk `.*` introduces and how it can cause our matcher to accidentally consume significantly more than we expect it to (and perform much slower while doing it!)

It's unfortunate we have more cases where we've already introduced these risks but I believe that's a good enough reason to be on alert to help spot these before they get merged. RegExp patterns can be hard enough, let alone with the complexity of trying to parse HTML…

---

In summary I think there's several ways to move this forward without throwing much out. Replacing the `.*` addresses the biggest issue. The point about associating a start tag with its end tag isn't a demand; it's an acknowledgement of how this stuff works unless we build far more complicated patterns (purely due to the recursive nature of HTML).

dmsnell reviewed (changes requested):
Marking "request changes" because I think we really want to avoid merging this RegExp pattern due to the potential for side-effects that could end up hard to diagnose.

draganescu commented:
I see this is really getting into a huge thing from a small PR that removed spaces in the incoming block markup, _for one block._ I would love to avoid turning it into a years long Trac ticket (joke). Let's make a summary:

1. The cover block and many other blocks use string operations on the HTML markup that is incoming to their server side render functions to insert various dynamic data.
2. @navigatrum discovered that in the case of the cover block it's easy to break the rendering by hand coding a cover block with some space between the block's markup tags. 
3. This PR solves that. It does not really introduce anything **new**, **systemic**, or with deep consequences to the future.

In my initial explorations of adding support for the featured image in the cover block I tried various ways of minimizing the amount of ad-hoc intervention at render time. However, the goals  of:

- keeping the block attributes clean and only contain up to date information, and
- having the featured image update dynamically when other parts of WP update it

... resulted in this situation: inserting a whole `img` tag at render time in a specific place in the block's markup. Ideally this is temporary and we can come up with a system of automatic dynamic replacement of tokens at render time for all blocks.

The larger discussions this PR provoked are:

1. We should have a programmatic way to build DOM when rendering blocks, but PHP's DOM is unreliable. What to do?
2. Hand coding block markup makes developing blocks unreliable. This **appeared** in the cover block, however raises the problem of expecting the markup to be in a shape and then getting it in a different shape because it is not from a save function but from hand written code. Should we have a cleanup system before the renderer gets the markup applied for all blocks?
3. Regexp operations on HTML are not great, the code that results is confusing and outside of comments we don't have much to offer for our future selves or future developers. I don't think this thread has much actionable stuff and it's easy to devolve into bike shedding.

Do y'all think it would be wise to open **separate issues** or discussion threads about the above, and on this PR decide if either one of the following is the path forward:

(a) Accept a regexp that removes spaces from the cover block incoming markup at render time. I would like to mention that the current form of the regexp is overly complex to make the code more readable, as much simpler forms existed in previous commits but were deemed too brief.
(b) Not do anything about it, close this PR. We should discuss - separately - what out expectation is in regard to hand coding block markup.
(c) Close this PR and separately explore ways to improve how we insert the featured image tag such that we don't depend on empty space anymore.

@navigatrum is a first time contributor here and the "ramifications of the implications of the situations" 😂  from the initial goal seem to be prone to be lost in the sea of PR comments, code comments, replies with specific people tagged and so on 😁 

I am the first to "blame" as I should have opened a separate discussion about the use of PHP's DOM, instead of asking here. But better late than never: (a) or (b) or (c) or something else on point that I can't think of right now?

dmsnell commented:
@draganescu I don't think we need to over-react. this PR has taken a wayward path but it's almost ready.

when I jumped in I was recommending and still recommending avoiding (a) because of the double-hit of being fragile plus hiding intent which will make it harder to fix the next time someone changes the format of the html. the current state of this PR doesn't do that so I don't think (a) needs further discussion.

for (b) and (c) - why close it? this PR fixes a real issue, no? or was the issue only there when manually modifying the HTML of a stored post? (vs. being created through the editor itself) regardless, this is a good change to make the cover block more reliable. it's good to avoid breaking things 😄

_for followup_ why can't we put an empty `<img />` in there and replace on the server? that we can do with a basic string search and replace.

azaozz commented:
> this is really getting into a huge thing from a small PR...

Yeah, sorry, my bad there for hijacking it. You're right, this is a simple PR to fix a simple problem. 

> Ideally this is temporary and we can come up with a system of automatic dynamic replacement of tokens at render time for all blocks.

Yes, big +1 for this. I know it sounds like "shortcodes" but hoping this time it can be done a lot more sensibly and avoid the numerous pitfalls the shortcodes have.

> think it would be wise to open separate issues or discussion...

Yes absolutely agree. Neither DOMDocument nor RegExp is good for parsing HTML in PHP. Would be best to choose another solution (the above?) and then remove all the similar RegExp once and for all.

azaozz commented on packages/block-library/src/cover/index.php:
```diff
@@ -60,10 +60,13 @@ class="wp-block-cover__image-background"
+		/* Inserts the featured image between the (1st) cover 'background' `span` and 'inner_container' `div`,
+		   and removes eventual withespace characters between the two (typically introduced at template level) */
+		$content = preg_replace(
+		    '/(<span\b[^>]*wp-block-cover__background[\s|"][^>]*>.*<\/span>)\s*(<div\b[^>]*wp-block-cover__inner-container[\s|"][^>]*>)/U',
```
Looking at the regex again, do you think the `span` tag matching can be removed? Doesn't seem needed as long as the opening `div` is matched. Even if there is a possibility to have the div without a span before it, thinking it would be good to insert the featured image before that div. This also removes the need to use `.*` and match white space between the nodes.

So the regex would be something like: `'/<div\b[^>]+wp-block-cover__inner-container(\s|")[^>]*>/U'`.
(Note I also changed `*` to `+` after the first `[^>]` as it has to match at least one char.)

navigatrum replied:
To me the logic of the block is:
the image goes between a `span` with class "wp-block-cover__background" followed by a div with class "wp-block-cover__inner-container". If you remove one or the other, no image. What if I want to put something inside the span which in the react code is empty? Good luck!

To you and @[dmsnell](https://github.com/dmsnell) since regexes are problematic for HTML parsing, and for a better maintainability, the php porting of the react logic should be reduced to: the featured image goes before a div with class "wp-block-cover__inner-container". Do what you want before that div and good luck! 

Or , dsmell variant: if we want to require both the `span` and the `div`, at that point we must require an empty span.

Now, just in case for @draganescu and @dmsnell  `'/<div\b[^>]+wp-block-cover__inner-container[\s|"][^>]*>/U'` by  @azaozz is acceptable, and given that 

```php
$inner_container_start = // above regex;

if ( 1 !== preg_match( $inner_container_start, $content, $matches, PREG_OFFSET_CAPTURE ) {
    // handle this, we can't find our div, maybe another plugin changed it
    return $content;
}

list( $fullmatch, $offset ) = $matches[0];
$content = substr( $content, 0, $offset ) . $image_tag . substr( $content, $offset );
```
shuold be acceptable by @dmsnell 
I'm sending the 4th PR to trim the coverblock (in a few hours)
(Sorry if a wasted your time)

draganescu replied:
@navigatrum no wasted time at all, quite the reverse it will be a fruitful conversation :)

dmsnell replied:
this all seems good, though I should rephrase, as I never wanted to suggest we try and trap the `<span>`. here's the difference:

 - finding the start of the `<div>` with a RegExp is normal and doable and _not problematic_ from a parsing standpoint. this is because _it isn't parsing HTML_ at that point. it's only parsing HTML tokens, or _lexing_ the HTML.
 - trying to identify the end of a specific `<span>` before the `<div>` _is_ parsing HTML and that's the problematic part with the RegExp and that's the kind of thing we want to avoid (and it's almost always the case to be suspicious when we find ourselves adding `.*`)

seems like a nuance or a small thing, but there's a profound impact. my comments were more like, "if you are going to match the `<span>` too here's what you should be aware of).

I still don't understand why we don't want to match on `class`, but that's not a big issue to me since `wp-block-cover__inner-container` is unlikely to appear outside of where we expect it. note though, this does match a few quirky possibilities:

 - `<div title="Demonstrating how to use the wp-block-cover__inner-container styles">`
 - `<div class="x-disregard-wp-block-cover__inner-container">`

<details><summary>Feel free to ignore matching these, but if you are curious…</summary>

```
"~<div [^>]*\bclass=(\"|')((?!\1).)*\b{$CLASS}\b((?!\1).)*\1( |>)~"
```

This is almost the same as the one you gave above in your last comment. The main differences are that this:
 - specifically looks for the `class` attribute
 - makes sure we're only examining the content inside the `class` attribute by matching on the leading quote and trailing quote, ensuring they are the same, and ensuring we don't scan past the matching one with `((?!\1).)*` though I think we could replace that with the simpler `[^'"]*` since I think quotes aren't allowed as class names.
 - requires that the class name we're searching for is not a part of a bigger string through the surrounding `\b` boundary markers (that's wrong, should be `[\"' ]` since the hyphen counts as a word boundary)
 - while we're not matching a closing HTML tag I'm still using `~` instead of `/` as the PCRE pattern delimiters so we wouldn't have to escape `/` as `\/`

with the simplification that would be this instead

```
"~<div [^>]*\bclass=([\"'])[^\"']*[\"' ]wp-block-cover__inner-container[\"' ][^\"']*\1~"
```

this is simpler than general HTML attribute parsing because of the rules of CSS class names. hopefully that fulfills your curiosity. feel free to use this or ask questions or ignore it.

draganescu commented:
@dmsnell on

> for (b) and (c) - why close it? 

If we don't want to strip whitespace as this PR proposes with a regexp there is no way forward for it. From your comment tho I think we're ok with how it evolved.

> this PR fixes a real issue, no? or was the issue only there when manually modifying the HTML of a stored post? 

This PR aims to fix **one issue for one block**: manually editing or creating the HTML for the cover block (say in a manually coded template part, or via code view in the editor) risks breaking featured image support because of unexpected whitespace.

> regardless, this is a good change to make the cover block more reliable. it's good to avoid breaking things 😄

+1 to that and the reason I support this. It's a very specific fix but a good one that does not tie us down to anything.

dmsnell commented:
@draganescu 

> If we don't want to strip whitespace as this PR proposes with a regexp there is no way forward for it.

not necessary in my opinion. I'd rephrase the problem, and if we wanted we could rename the PR but the purpose is clear: fix the broken cover block image injection.

"Avoid whitespace-sensitivity when injecting cover block featured image" is one possible rephrasing of the problem. At this point I don't think it matters if we rename the PR or not because the primary goal is still the same.

draganescu commented:
@youknowriad as we've [discussed](https://github.com/WordPress/gutenberg/pull/39658#discussion_r836301775) while implementing the featured image for cover block it's best to keep block attributes clean (e.g. featured image to just be a flag) and the block markup clean (e.g. not containing superfluous content which would not be stripped on export, etc.).

I wonder though (in line with what @azaozz asked as well) in light of the problems this PR tries to solve - inserting html at render time - would it be acceptable to place a token in the form of an HTML comment `<!--wp-featured-image-token-->` if `useFeaturedImage` is `true` via the `save` function, and replace that on render?

There will be situations where the token will remain there but since it's "only" an HTML comment I think it's not that bad. That would solve the whole discussion here.

scruffian commented:
IMO a simple regex is the best solution for this particular problem right now. We should look in to a more comprehensive solution that we can use in all the many places where we current do regex replaces, but that should not stop this from being merged, so that we can fix the bug.

adamziel commented:
The following way forward makes sense to me:

1. Go with the regex for now (perhaps more robust one to reflect the discussion above).
2. Acknowledge it won't perfectly cover all cases.
3. Start another PR to explore a full solution. That could mean adding a proper DOM parser (not DOMDocument) as a composer dependency and then leaning on it.

draganescu commented:
It seems we can proceed here with a regexp and we have two [suggestions](https://github.com/WordPress/gutenberg/pull/40859#discussion_r886349788) from @dmsnell 👍🏻  

1. `~<div [^>]*\bclass=(\"|')((?!\1).)*\b{$CLASS}\b((?!\1).)*\1( |>)~` or, more simply,  
2. `~<div [^>]*\bclass=([\"'])[^\"']*[\"' ]wp-block-cover__inner-container[\"' ][^\"']*\1~`

and [the one](https://github.com/WordPress/gutenberg/pull/40859#discussion_r884068877) from @navigatrum :

3. `'/<div\b[^>]+wp-block-cover__inner-container[\s|"][^>]*>/U'` 

... which are very similar.

I think (1) is hard to follow, (2) and (3) are better, with (2) being even safer with the match on the "class" attribute name.

 @navigatrum let's tweak the regex one last time and then this will land 👏🏻 You still up for it?

navigatrum committed:
edit cover block regex for featured image #4

For the block layout see: https://github.com/WordPress/gutenberg/blob/trunk/packages/block-library/src/cover/edit/index.js#L327-L367
For the HTML tags capturing by class name see: https://github.com/WordPress/gutenberg/blob/51db4bf888e6b18cf9d18266108114d61070f3ad/lib/block-supports/layout.php#L232

adamziel commented on packages/block-library/src/cover/index.php:
```diff
@@ -60,11 +60,13 @@ class="wp-block-cover__image-background"
-		);
+		/* Inserts the featured image between the (1st) cover 'background' `span` and 'inner_container' `div`,
+		   and removes eventual withespace characters between the two (typically introduced at template level) */
+		$inner_container_start = /<div\b[^>]+wp-block-cover__inner-container[\s|"][^>]*>/U;
```
@navigatrum this change broke the unit tests. It's a syntax error, regexps in PHP must be enclosed in quotes like this:

```suggestion
		$inner_container_start = '/<div\b[^>]+wp-block-cover__inner-container[\s|"][^>]*>/U';
```

adamziel replied:
I went ahead and committed that suggestion to see if it's going to fix all the tests 🤞

adamziel committed:
Add quotes around the regexp

dmsnell reviewed (approved):
Pending the tests passing and this working as we expect it to, I approve 👍 

Thanks for persevering with this!

scruffian commented:
@navigatrum there's a PHP linting error on this, are you able to address it? Thanks!

navigatrum commented:
> @navigatrum there's a PHP linting error on this, are you able to address it? Thanks!

Wasn't it quotes around regex solved by admziel commit?

scruffian commented:
> Wasn't it quotes around regex solved by admziel commit?

There's another one with a missing `;`

navigatrum committed:
close bracket ad reformat comment

scruffian commented:
Sorry I'm still seeing these errors:

```
FOUND 1 ERROR AND 4 WARNINGS AFFECTING 4 LINES
----------------------------------------------------------------------
 33 | WARNING | [ ] Unused variable $image.
    |         |     (VariableAnalysis.CodeAnalysis.VariableAnalysis.UnusedVariable)
 41 | WARNING | [ ] Unused variable $fullmatch.
    |         |     (VariableAnalysis.CodeAnalysis.VariableAnalysis.UnusedVariable)
 42 | WARNING | [x] Equals sign not aligned with surrounding
    |         |     assignments; expected 20 spaces but found 1
    |         |     space
    |         |     (Generic.Formatting.MultipleStatementAlignment.NotSameWarning)
 42 | WARNING | [ ] Variable $image_tag is undefined.
    |         |     (VariableAnalysis.CodeAnalysis.VariableAnalysis.UndefinedVariable)
 43 | ERROR   | [x] Blank line found after control structure
    |         |     (WordPress.WhiteSpace.ControlStructureSpacing.BlankLineAfterEnd)
----------------------------------------------------------------------
```

navigatrum committed:
retry to fix

adamziel commented on packages/block-library/src/cover/index.php:
```diff
@@ -60,11 +60,15 @@ class="wp-block-cover__image-background"
+		 */
+		$inner_container_start = '/<div\b[^>]+wp-block-cover__inner-container[\s|"][^>]*>/U';
+		if ( 1 === preg_match( $inner_container_start, $content, $matches, PREG_OFFSET_CAPTURE ) ) {
+			$offset = $matches[0][1];
```
```suggestion
			$offset  = $matches[0][1];
```

adamziel commented on packages/block-library/src/cover/index.php:
```diff
@@ -60,11 +60,15 @@ class="wp-block-cover__image-background"
+			$offset = $matches[0][1];
+			$content = substr( $content, 0, $offset ) . $image . substr( $content, $offset );
+		}
 
```
```suggestion
```

adamziel committed:
Update packages/block-library/src/cover/index.php

adamziel committed:
Update packages/block-library/src/cover/index.php

adamziel committed:
Merge branch 'trunk' into edit-cover-block-featured-image

adamziel commented:
The failing navigation E2E tests are flaky and unrelated to this PR. Let's ship it!

draganescu commented:
Thanks:
- @navigatrum on the work and effort to see this through
- @dmsnell for the great insights
- everyone else for the participation \o/
- @adamziel for the final button push 😁

adamziel commented:
I just cherry-picked this PR to the wp/6.0 branch to get it included in the next release: c92a83a620
